packages/core/src/tools/instrumentMethod.ts

```diff
@@ -36,135 +35,322 @@ type PostCallCallback<TARGET extends { [key: string]: any }, METHOD extends keyo
+      // If this observer is still attached to this hook, remove it, and
+      // recompute shared hook state.
+      const index = observers.indexOf(observer)
+      if (index > -1) {
+        observers.splice(index, 1)
+        anyRequireHandlingStack = observers.some((item) => item.requiresHandlingStack)
+      }
+
+      // If we still have observers attached to this hook, leave it in place.
+      if (observers.length > 0) {
+        return
+      }
+
+      // We've removed the last observer. If the underlying method has
+      // not been overridden in the meantime, remove the hook.
+      if (targetPrototype[method] === hookFunction) {
         targetPrototype[method] = original
+        delete objectHooks.methods[method]
       }
     },
+    getObservers() {
+      return observers.concat()
+    },
   }
+
+  objectHooks.methods[method] = methodHook
+  targetPrototype[method] = hookFunction
+  return methodHook
 }
 
-export function instrumentSetter<TARGET extends { [key: string]: any }, PROPERTY extends keyof TARGET>(
+function getOrInstallPropertyHook<TARGET extends { [key: string]: any }, PROPERTY extends keyof TARGET>(
+  objectHooks: ObjectHooks<TARGET>,
   targetPrototype: TARGET,
-  property: PROPERTY,
-  after: (target: TARGET, value: TARGET[PROPERTY]) => void
-) {
-  const originalDescriptor = Object.getOwnPropertyDescriptor(targetPrototype, property)
-  if (!originalDescriptor || !originalDescriptor.set || !originalDescriptor.configurable) {
-    return { stop: noop }
+  property: PROPERTY
+): PropertyHook<TARGET, PROPERTY> | undefined {
+  const existingPropertyHook: PropertyHook<TARGET, PROPERTY> | undefined = objectHooks.properties[property]
+  if (existingPropertyHook) {
+    return existingPropertyHook
+  }
+
+  const original = Object.getOwnPropertyDescriptor(targetPrototype, property)
+  if (!original || !original.set || !original.configurable) {
+    return undefined
   }
 
-  const stoppedInstrumentation = noop
-  let instrumentation = (target: TARGET, value: TARGET[PROPERTY]) => {
-    // put hooked setter into event loop to avoid of set latency
-    setTimeout(() => {
-      if (instrumentation !== stoppedInstrumentation) {
-        after(target, value)
+  const observers: Array<PropertyObserver<TARGET, PROPERTY>> = []
+
+  const hookDescriptor: PropertyDescriptor = {
+    set(this: TARGET, value: TARGET[PROPERTY]): void {
+      original.set!.call(this, value)
+
+      if (observers.length === 0) {
+        return
       }
-    }, 0)
+
+      // Invoke observers in separate microtasks to avoid setter latency.
+      //
+      // Note that we deliberately don't capture the active observers here; as a
+      // design decision, we've chosen to allow changes in the observer list to
+      // take effect immediately, even if they occur between the time a setter is called
+      // and the time at which its observers are invoked.
+      // Context: https://github.com/DataDog/browser-sdk/pull/2598
```

Comment:
@amortemousque I figured I should ping you on this just to verify that nothing has changed. We have a test for this, so it's pretty clear that it was intended, but I know things change over time and I didn't want to assume that this was set in stone just because there's a test.

I may have been too conservative about changing that, but I may also have been too cavalier about switching to promises here. The old code used `setTimeout()`, so the property observers would run in a separate macrotask, but now they'll run in the same macrotask, just in a separate microtask. My guess is that that's preferable, since it reduces the likelihood of the value getting stale or us reacting to it too late, but if we do really expensive things in these property observer callbacks it might be that `setTimeout()` is better!

(Reposting this comment; I pushed a small change to this PR which changed the indentation of this code, so the original version got marked as "Outdated".)